frame/balances/src/lib.rs

```diff
@@ -721,14 +731,20 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
 				}
 			}
 		});
-		Locks::<T, I>::insert(who, locks);
-	}
-}
 
-impl<T: Trait<I>, I: Instance> OnReapAccount<T::AccountId> for Module<T, I> {
-	fn on_reap_account(who: &T::AccountId) {
-		Locks::<T, I>::remove(who);
-		Account::<T, I>::remove(who);
+		let existed = Locks::<T, I>::contains_key(who);
+		if locks.is_empty() {
+			Locks::<T, I>::remove(who);
+			if existed {
+				// TODO: use Locks::<T, I>::hashed_key
```

Comment:
```suggestion
				// TODO: use Locks::<T, I>::hashed_key
				// https://github.com/paritytech/substrate/issues/4969
```

Reply:
@shawntabrizi Hey, did you mean using `hashed_key` like that to detect a change more efficiently?

```rust
...
		let old_key = Locks::<T, I>::hashed_key_for(who);
		if locks.is_empty() {
			Locks::<T, I>::remove(who);
			// account existed
			if old_key != Locks::<T, I>::hashed_key_for(who) {
				// TODO: use Locks::<T, I>::hashed_key
				// https://github.com/paritytech/substrate/issues/4969
				system::Pallet::<T>::dec_consumers(who);
			}
		} else {
			Locks::<T, I>::insert(who, bounded_locks);
			// account didn't exist
			if old_key == Locks::<T, I>::hashed_key_for(who) {
				if system::Pallet::<T>::inc_consumers(who).is_err() {
					// No providers for the locks. This is impossible under normal circumstances
					// since the funds that are under the lock will themselves be stored in the
					// account and therefore will need a reference.
```